src/modules/Game.class.js

```diff
@@ -20,25 +27,321 @@ class Game {
+          // if not equal, start = second
+          start = second;
+          continue;
+        }
+      }
+    }
+
+    for (let i = 0; i < 4; i++) {
+      for (let j = 0; j < 4; j++) {
+        if (this.state[i][j] !== 0) {
+          const temp = this.state[i][j];
+
+          this.state[i][j] = 0;
+
+          const idx = this.state[i].findIndex((el) => el === 0);
+
+          this.state[i][idx] = temp;
+
+          if (j !== idx) {
+            moveSuccess = true;
+          }
+        }
+      }
+    }
+
+    return moveSuccess;
+  }
+
+  moveRight() {
+    let moveSuccess = false;
+
+    for (let row = 0; row < 4; row++) {
+      let start = 3;
+
+      while (start > 0) {
+        // find first non zero element from right
+        let first = -1;
+
+        for (let i = start; i > 0; i--) {
+          if (this.state[row][i] !== 0) {
+            first = i;
+            break;
+          }
+        }
+
+        // find next non zero element
+        let second = -1;
+
+        for (let i = first - 1; i >= 0; i--) {
+          if (this.state[row][i] !== 0) {
+            second = i;
+            break;
+          }
+        }
+
+        if (first === -1 || second === -1) {
+          break;
+        }
+
+        // if equal merge, next round start = second + 1
+        if (this.state[row][first] === this.state[row][second]) {
+          this.state[row][first] *= 2;
+          this.state[row][second] = 0;
+          moveSuccess = true;
+          this.addScore(this.state[row][first]);
+          start = second - 1;
+          continue;
+        } else {
+          // if not equal, start = second
+          start = second;
+          continue;
+        }
+      }
+    }
+
+    for (let i = 0; i < 4; i++) {
+      for (let j = 3; j >= 0; j--) {
+        if (this.state[i][j] !== 0) {
+          const temp = this.state[i][j];
+
+          this.state[i][j] = 0;
+
+          let idx = null;
+
+          // find index of the right-most zero
+          for (let q = 3; q >= 0; q--) {
+            if (this.state[i][q] === 0) {
+              idx = q;
+              break;
+            }
+          }
+
+          this.state[i][idx] = temp;
+
+          if (j !== idx) {
+            moveSuccess = true;
+          }
+        }
+      }
+    }
+
+    return moveSuccess;
+  }
+
+  moveUp() {
+    let moveSuccess = false;
+
+    for (let col = 0; col < 4; col++) {
+      let start = 0;
+
+      while (start < 4) {
+        // find first non zero element from right
+        let first = -1;
+
+        for (let i = start; i < 3; i++) {
+          if (this.state[i][col] !== 0) {
+            first = i;
+            break;
+          }
+        }
+
+        // find next non zero element
+        let second = -1;
+
+        for (let i = first + 1; i < 4; i++) {
+          if (this.state[i][col] !== 0) {
+            second = i;
+            break;
+          }
+        }
+
+        if (first === -1 || second === -1) {
+          break;
+        }
```

Comment:
This logic is now similar in every method, try saving it in a helper function and using it everywhere